scylla-cql/src/types/deserialize/mod.rs

```diff
@@ -0,0 +1,166 @@
+pub mod row;
+pub mod value;
+
+use bytes::Bytes;
+
+use crate::frame::frame_errors::ParseError;
+use crate::frame::types;
+
+/// A reference to a part of the frame.
+#[derive(Clone, Copy, Debug)]
+pub struct FrameSlice<'frame> {
+    // The actual subslice represented by this FrameSlice.
+    frame_subslice: &'frame [u8],
+
+    // Reference to the original Bytes object that this FrameSlice is derived
+    // from. It is used to convert the `mem` slice into a fully blown Bytes
+    // object via Bytes::slice_ref method.
+    original_frame: &'frame Bytes,
+}
+
+static EMPTY_BYTES: Bytes = Bytes::new();
+
+impl<'frame> FrameSlice<'frame> {
+    /// Creates a new FrameSlice from a reference of a Bytes object.
+    ///
+    /// This method is exposed to allow writing deserialization tests
+    /// for custom types.
+    #[inline]
+    pub fn new(frame: &'frame Bytes) -> Self {
+        Self {
+            frame_subslice: frame,
+            original_frame: frame,
+        }
+    }
+
+    /// Creates a new FrameSlice that refers to a subslice of a given Bytes object.
+    #[inline]
+    pub fn new_subslice(mem: &'frame [u8], frame: &'frame Bytes) -> Self {
+        Self {
+            frame_subslice: mem,
+            original_frame: frame,
+        }
+    }
+
+    /// Creates an empty FrameSlice.
+    #[inline]
+    pub fn new_empty() -> Self {
+        Self {
+            frame_subslice: &EMPTY_BYTES,
+            original_frame: &EMPTY_BYTES,
+        }
+    }
+
+    /// Returns a reference to the slice.
+    #[inline]
+    pub fn as_slice(&self) -> &'frame [u8] {
+        self.frame_subslice
+    }
+
+    /// Returns `true` if the slice has length of 0.
+    #[inline]
+    pub fn is_empty(&self) -> bool {
+        self.frame_subslice.is_empty()
+    }
+
+    /// Returns a reference to the Bytes object which encompasses the slice.
+    ///
+    /// The Bytes object will usually be larger than the slice returned by
+    /// [FrameSlice::as_slice]. If you wish to obtain a new Bytes object that
+    /// points only to the subslice represented by the FrameSlice object,
+    /// see [FrameSlice::to_bytes].
+    #[inline]
+    pub fn as_bytes_ref(&self) -> &'frame Bytes {
+        self.original_frame
+    }
+
+    /// Returns a new Bytes object which is a subslice of the original slice
+    /// object.
+    #[inline]
+    pub fn to_bytes(&self) -> Bytes {
+        self.original_frame.slice_ref(self.frame_subslice)
+    }
```

Comment:
### Why we need a slice
The reason why we need to store a slice is that we want a lifetime that is longer than one obtained when coercing `Bytes` to a slice in the body of a function. That is, we want to allow deserializing types that borrow from the frame, which resides in `QueryResult`.
Consider a function with the signature:
```rust
fn fun(b: Bytes) { ... }
```
This function cannot return a type that borrows from the frame, because any slice created from `b` inside `fun` cannot escape `fun`.
Conversely, if a function has signature:
```rust
fn fun(s: &'frame [u8]) { ... }
```
then it can happily return types with lifetime `'frame`.

### Why we need the full frame?
We don't. We only need to be able to return `Bytes` encompassing our subslice. However, the design choice was made to only store a reference to the original `Bytes` object residing in `QueryResult`, so that we avoid cloning `Bytes` when performing subslicing on `FrameSlice`. We delay the `Bytes` cloning, normally a moderately expensive operation involving cloning an `Arc`, up until it is really needed.

### Why not different design?
- why not a `&'frame [u8]` only? Because we won't to enable deserializing types containing owned `Bytes`, too.
- why not a `Bytes` only? Because we need to propagate the `'frame` lifetime.
- why not a `&'frame Bytes` only? Because we want to somehow represent subslices, and subslicing `&'frame Bytes` return `Bytes`, not `&'frame Bytes`.

Is the above convincing enough? Please hint me which of the above would you like to be put in the comment explanation.